The job **`aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/22267426339/job/64416143800) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/64416143800) [(plain)](https://github.com/rust-lang/rust/commit/5a68e1a7a4c7e95d0e6c36eba36c94f137611330/checks/64416143800/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
/dev/disk1s2     500Mi    20Ki   495Mi     1%       0  5.1M    0%   /System/Volumes/xarts
/dev/disk1s1     500Mi   116Ki   495Mi     1%      25  5.1M    0%   /System/Volumes/iSCPreboot
/dev/disk1s3     500Mi   224Ki   495Mi     1%      18  5.1M    0%   /System/Volumes/Hardware
/dev/disk3s5     320Gi   256Gi    45Gi    85%    2.2M  474M    0%   /System/Volumes/Data
/dev/disk5s1     4.0Mi   676Ki   3.0Mi    19%      18   31k    0%   /System/Library/AssetsV2/com_apple_MobileAsset_PKITrustStore/purpose_auto/6dd55b0d06633a00de6f57ccb910a66a5ba2409a.asset/.AssetData
map auto_home      0Bi     0Bi     0Bi   100%       0     0     -   /System/Volumes/Data/home
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
shell: /bin/bash --noprofile --norc -e -o pipefail {0}
---
##[endgroup]
==> Fetching downloads for: tidy-html5
✔︎ Bottle Manifest tidy-html5 (5.8.0)
✔︎ Bottle tidy-html5 (5.8.0)
==> Pouring tidy-html5--5.8.0.arm64_sequoia.bottle.tar.gz
🍺  /opt/homebrew/Cellar/tidy-html5/5.8.0: 15 files, 3.2MB
##[group]Run src/ci/scripts/install-wix.sh
src/ci/scripts/install-wix.sh
shell: /bin/bash --noprofile --norc -e -o pipefail {0}
---
hw.optional.arm.SME_I16I32: 0
hw.optional.arm.FEAT_SME_F64F64: 0
hw.optional.arm.FEAT_SME_I16I64: 0
hw.optional.arm.FP_SyncExceptions: 1
hw.optional.arm.caps: 292171059125284863
hw.optional.arm.sme_max_svl_b: 0
hw.optional.floatingpoint: 1
hw.optional.neon: 1
hw.optional.neon_hpfp: 1
hw.optional.neon_fp16: 1
hw.optional.armv8_crc32: 1
---
fmt: skipped untracked file AWSCLIV2.pkg
fmt: checked 6707 files
tidy check
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy [style (src)]: /Users/runner/work/rust/rust/src/bootstrap/src/core/builder/tests.rs: too many lines (3014) (add `// ignore-tidy-filelength` to the file to suppress this error)
tidy [style (src)]: FAIL
tidy: The following check failed: style (src)
Command `/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage1-tools-bin/rust-tidy --root-path=/Users/runner/work/rust/rust --cargo-path=/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0/bin/cargo --output-dir=/Users/runner/work/rust/rust/build --concurrency=3 --npm-path=/Users/runner/.yarn/bin/yarn` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1612:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1379:29

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `--stage 2 test --host=aarch64-apple-darwin --target=aarch64-apple-darwin`
Build completed unsuccessfully in 0:02:29
  local time: Sun Feb 22 00:48:21 UTC 2026
  network time: Sun, 22 Feb 2026 00:48:22 GMT
##[error]Process completed with exit code 1.
---
/dev/disk1s2     500Mi    20Ki   495Mi     1%       0  5.1M    0%   /System/Volumes/xarts
/dev/disk1s1     500Mi   116Ki   495Mi     1%      25  5.1M    0%   /System/Volumes/iSCPreboot
/dev/disk1s3     500Mi   224Ki   495Mi     1%      18  5.1M    0%   /System/Volumes/Hardware
/dev/disk3s5     320Gi   264Gi    37Gi    88%    2.7M  385M    1%   /System/Volumes/Data
/dev/disk5s1     4.0Mi   676Ki   3.0Mi    19%      18   31k    0%   /System/Library/AssetsV2/com_apple_MobileAsset_PKITrustStore/purpose_auto/6dd55b0d06633a00de6f57ccb910a66a5ba2409a.asset/.AssetData
map auto_home      0Bi     0Bi     0Bi   100%       0     0     -   /System/Volumes/Data/home
Post job cleanup.
[command]/opt/homebrew/bin/git version
git version 2.53.0
Copying '/Users/runner/.gitconfig' to '/Users/runner/work/_temp/d6a5ce5e-90d5-492b-b90d-601c4598362a/.gitconfig'
```

</details>

--- a/src/bootstrap/src/core/build_steps/test.rs
+++ b/src/bootstrap/src/core/build_steps/test.rs
@@ -805,6 +805,20 @@ impl Step for CargoMiri {
             &[],
         );
 
+        // If we are testing stage 2+ cargo miri, make sure that it works with the in-tree cargo.
+        // We want to do this *somewhere* to ensure that Miri + nightly cargo actually works.
+        if stage >= 2 {
+            let built_cargo = builder
+                .ensure(tool::Cargo::from_build_compiler(
+                    // Build stage 1 cargo here, we don't need it to be built in any special way,
+                    // just that it is built from in-tree sources.
+                    builder.compiler(0, builder.host_target),
+                    builder.host_target,
+                ))
+                .tool_path;
+            cargo.env("CARGO", built_cargo);
+        }
+
         // We're not using `prepare_cargo_test` so we have to do this ourselves.
         // (We're not using that as the test-cargo-miri crate is not known to bootstrap.)
         match builder.doc_tests {

--- a/src/bootstrap/src/core/builder/tests.rs
+++ b/src/bootstrap/src/core/builder/tests.rs
@@ -3079,6 +3079,45 @@ mod snapshot {
         ");
     }
 
+    #[test]
+    fn cargo_miri_stage_1() {
+        let ctx = TestCtx::new();
+        insta::assert_snapshot!(
+            ctx.config("test")
+                .args(&["cargo-miri"])
+                .stage(1)
+                .get_steps()
+                .render(), @"
+        [build] llvm <host>
+        [build] rustc 0 <host> -> rustc 1 <host>
+        [build] rustc 0 <host> -> miri 1 <host>
+        [build] rustc 0 <host> -> cargo-miri 1 <host>
+        [build] rustdoc 1 <host>
+        [build] rustc 1 <host> -> std 1 <host>
+        ");
+    }
+
+    #[test]
+    fn cargo_miri_stage_2() {
+        let ctx = TestCtx::new();
+        insta::assert_snapshot!(
+            ctx.config("test")
+                .args(&["cargo-miri"])
+                .stage(2)
+                .get_steps()
+                .render(), @"
+        [build] llvm <host>
+        [build] rustc 0 <host> -> rustc 1 <host>
+        [build] rustc 1 <host> -> std 1 <host>
+        [build] rustc 1 <host> -> rustc 2 <host>
+        [build] rustc 1 <host> -> miri 2 <host>
+        [build] rustc 1 <host> -> cargo-miri 2 <host>
+        [build] rustdoc 2 <host>
+        [build] rustc 2 <host> -> std 2 <host>
+        [build] rustc 0 <host> -> cargo 1 <host>
+        ");
+    }
+
     // Check that `x run miri --target FOO` actually builds miri for the host.
     #[test]
     fn run_miri() {

--- a/src/bootstrap/src/core/builder/tests.rs
+++ b/src/bootstrap/src/core/builder/tests.rs
@@ -1,3 +1,4 @@
+// ignore-tidy-filelength
 use std::env::VarError;
 use std::{panic, thread};
 

